docker/docker_client.go

```diff
@@ -244,37 +248,96 @@ func newDockerClient(sys *types.SystemContext, registry, reference string) (*doc
 	if err := tlsclientconfig.SetupCertificates(certDir, tr.TLSClientConfig); err != nil {
 		return nil, err
 	}
+	tr.TLSClientConfig.InsecureSkipVerify = sysContextSkipVerify(sys)
+
+	// Find the registry
+	reg, err := sysregistriesv2.FindRegistry(sys, reference)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+	if reg != nil {
+		// Retrieve the docker client from the found registry
+		return dockerClientFromSysRegistry(ctx, sys, reg, tr), nil
+	}
 
+	// No registry found within the configuration, fallback to the provided
+	// one
+	return &dockerClient{
+		registry:              registry,
+		sys:                   sys,
+		client:                &http.Client{Transport: tr},
+		insecureSkipTLSVerify: tr.TLSClientConfig.InsecureSkipVerify,
+	}, nil
+}
+
+func sysContextSkipVerify(sys *types.SystemContext) bool {
 	// Check if TLS verification shall be skipped (default=false) which can
 	// either be specified in the sysregistriesv2 configuration or via the
 	// SystemContext, whereas the SystemContext is prioritized.
-	skipVerify := false
 	if sys != nil && sys.DockerInsecureSkipTLSVerify != types.OptionalBoolUndefined {
 		// Only use the SystemContext if the actual value is defined.
-		skipVerify = sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
-	} else {
-		reg, err := sysregistriesv2.FindRegistry(sys, reference)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading registries")
+		return sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
+	}
+	return false
+}
+
+// dockerClientFromSysRegistry returns a `*dockerClient` for the provided
+// sysregistriesv2. If the registry contains any mirrors, it will be tried
+// to detect their properties. The mirror will be used if the property
+// detection succeeds. If not, the fallback registry URL will be used instead.
+func dockerClientFromSysRegistry(
+	ctx context.Context,
+	sys *types.SystemContext,
+	registry *sysregistriesv2.Registry,
+	transport *http.Transport,
```

Comment:
Can we do a s/transport/httpTransport/ to avoid confusing with `cotainers/image/transports`?

Reply:
> Can we do a s/transport/httpTransport/ to avoid confusing with `cotainers/image/transports`?

Sure, with my new proposed structure this will go away.